File: libnd4j/include/ops/declarable/generic/random/multinomial.cpp
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

//
// @author Oleh Semeniv (oleg.semeniv@gmail.com)
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_random_multinomial)

#include <ops/declarable/CustomOperations.h>
#include <helpers/RandomLauncher.h>
#include <ops/declarable/helpers/random.h>

namespace sd {
    namespace ops {
        ///////////////////////
        /**
         * multinomial (categorical) random generator
         * takes 2D ndarray with logits with shape [batch_size (N), num_classes (K)]
         * and array with one scalar value of samples number, number of independent samples to draw for each experiment 1,N.
         * represents the unnormalized log-probabilities for all classes.
         * Int arguments: 0 - optional argument, corresponds to dimension with batch_size
         * Int arguments: 1 - optional argument, integer type to use for the output. Default int64.
         */
         // used https://en.wikipedia.org/wiki/Categorical_distribution
         // methods: gumbel trick + softmax + argmax
        CUSTOM_OP_IMPL(random_multinomial, 2, 1, false, 0, 0) {
            
            auto input = INPUT_VARIABLE(0);
            auto output = OUTPUT_NULLIFIED(0);
            auto inputSamples = INPUT_VARIABLE(1);


            REQUIRE_TRUE(!input->isEmpty(), 0, "RANDOM_MULTINOMIAL OP: Have to be provided at least one logits. ");

            REQUIRE_TRUE(inputSamples->lengthOf() == 1, 0, "RANDOM_MULTINOMIAL OP: Have to be specified at least one sample,"
                " but got no argumets instead.");
            
            Nd4jLong numOfSamples = static_cast<Nd4jLong>(inputSamples->e<int>(0));
            // do nothing if number of samples = 0
            if (0 == numOfSamples)
                return Status::OK();

            REQUIRE_TRUE(numOfSamples > 0, 0, "RANDOM_MULTINOMIAL OP: Number of samples should be greater then 0, got %i. ", numOfSamples);

            const int rank = input->rankOf();
            REQUIRE_TRUE(rank == 2, 0, "RANDOM_MULTINOMIAL OP: Logits should be a matrix with rank = 2, but got instead rank = %i.", rank);

            const int argSize = block.getIArguments()->size();
            const int dimC = argSize > 0 ? (INT_ARG(0) >= 0 ? INT_ARG(0) : INT_ARG(0) + rank) : rank - 1;

            auto dimA = (0 == dimC) ? 1 : 0;
            if (1 == input->sizeAt(dimA)) {
                *output = 0;
                return Status::OK();
            }

            auto rng = block.randomGenerator();
            helpers::fillRandomMultiNomial(block.launchContext(), rng, *input, *output, numOfSamples, dimC);
            return Status::OK();
        }


        DECLARE_SHAPE_FN(random_multinomial) {

            auto input = INPUT_VARIABLE(0);
            auto inputSamples = INPUT_VARIABLE(1);

            REQUIRE_TRUE(inputSamples->lengthOf() == 1, 0, "RANDOM_MULTINOMIAL OP: Have to be specified at least one sample,"
                " but got no argumets instead.");

            Nd4jLong numOfSamples = static_cast<Nd4jLong>(inputSamples->e<int>(0));

            REQUIRE_TRUE(numOfSamples > 0, 0, "RANDOM_MULTINOMIAL OP: Number of samples should be greater then 0, got %i. ", numOfSamples);

            const int rank = input->rankOf();
            REQUIRE_TRUE(rank == 2, 0, "RANDOM_MULTINOMIAL OP: Logits should be a matrix with rank = 2, but got instead rank = %i.", rank);

            const int argSize = block.getIArguments()->size();
            const int dimC = argSize > 0 ? (INT_ARG(0) >= 0 ? INT_ARG(0) : INT_ARG(0) + rank) : rank - 1;
           
            auto nShape = input->getShapeAsVector();
            auto dimA = (0 == dimC) ? 1 : 0;
            nShape[dimA] = numOfSamples;

            DataType nType = (argSize > 1) ? ( INT_ARG(1) >= 0 ? static_cast<DataType>(INT_ARG(1)) : sd::DataType::INT64) : sd::DataType::INT64;
            return SHAPELIST(ConstantShapeHelper::getInstance().createShapeInfo(nType, input->ordering(), nShape));
        }
        
        DECLARE_TYPES(random_multinomial) {
            getOpDescriptor()
                    ->setAllowedInputTypes(0, { ALL_FLOATS, ALL_INTS })
                    ->setAllowedInputTypes(1, { sd::DataType::INT32 })
                    ->setAllowedOutputTypes(0, { ALL_INDICES });
        }
    }
}

#endif

File: nd4j/samediff-import/samediff-import-onnx/src/main/kotlin/org/nd4j/samediff/frameworkimport/onnx/definitions/implementations/Multinomial.kt
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */
package org.nd4j.samediff.frameworkimport.onnx.definitions.implementations

import org.nd4j.autodiff.samediff.SDVariable
import org.nd4j.autodiff.samediff.SameDiff
import org.nd4j.autodiff.samediff.internal.SameDiffOp
import org.nd4j.linalg.api.buffer.DataType
import org.nd4j.samediff.frameworkimport.ImportGraph
import org.nd4j.samediff.frameworkimport.hooks.PreImportHook
import org.nd4j.samediff.frameworkimport.hooks.annotations.PreHookRule
import org.nd4j.samediff.frameworkimport.registry.OpMappingRegistry
import org.nd4j.shade.protobuf.GeneratedMessageV3
import org.nd4j.shade.protobuf.ProtocolMessageEnum

/**
 * Implementation of ONNX Multinomial operation.
 *
 * ONNX Multinomial spec: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Multinomial
 *
 * Generate random samples from a multinomial distribution.
 *
 * Inputs:
 * - input: Input tensor with unnormalized log-probabilities of shape [batch_size, class_size]
 *
 * Attributes:
 * - dtype: Output data type (default: INT32)
 * - sample_size: Number of samples to draw per batch (default: 1)
 * - seed: Random seed (optional)
 *
 * Output:
 * - output: Tensor of shape [batch_size, sample_size] with sampled indices
 *
 * @author Adam Gibson
 */
@PreHookRule(nodeNames = [], opNames = ["Multinomial"], frameworkName = "onnx")
class Multinomial : PreImportHook {

    override fun doImport(
        sd: SameDiff,
        attributes: Map<String, Any>,
        outputNames: List<String>,
        op: SameDiffOp,
        mappingRegistry: OpMappingRegistry<GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, ProtocolMessageEnum, GeneratedMessageV3, GeneratedMessageV3>,
        importGraph: ImportGraph<GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, ProtocolMessageEnum>,
        dynamicVariables: Map<String, GeneratedMessageV3>
    ): Map<String, List<SDVariable>> {
        val opName = op.name

        // Get input (unnormalized log-probabilities)
        val input = sd.getVariable(op.inputsToOp[0])

        // Get attributes
        val sampleSize = (attributes.getOrDefault("sample_size", 1L) as Number).toInt()

        // Get dtype
        val dtypeAttr = attributes["dtype"]
        val dtype = if (dtypeAttr != null) {
            val dtypeInt = (dtypeAttr as Number).toInt()
            when (dtypeInt) {
                6 -> DataType.INT32
                7 -> DataType.INT64
                else -> DataType.INT32
            }
        } else {
            DataType.INT32
        }

        // Convert log-probabilities to probabilities using softmax
        val probs = sd.nn.softmax("${opName}_probs", input, -1)

        // Multinomial sampling implementation using Gumbel trick:
        // argmax(log(probs) + Gumbel(0, 1)) gives samples from categorical distribution
        // For multiple samples, we repeat this process
        val logProbs = sd.math.log("${opName}_logprobs", probs.add(1e-10))

        // Generate Gumbel noise: -log(-log(U)) where U ~ Uniform(0, 1)
        val inputShape = probs.shape ?: throw IllegalStateException("Multinomial requires static input shape")
        val batchSize = inputShape[0]
        val classSize = inputShape[1]

        // For now, implement single sample (sample_size=1)
        // Generate uniform random
        val uniform = sd.random.uniform("${opName}_uniform", 1e-10, 1.0, DataType.FLOAT, batchSize, classSize)
        val gumbel = sd.math.neg(sd.math.log(sd.math.neg(sd.math.log("${opName}_log1", uniform))))

        // Add gumbel noise and take argmax
        val perturbed = sd.math.add("${opName}_perturbed", logProbs, gumbel)
        val samples = sd.argmax("${opName}_argmax", perturbed, false, -1L)

        val output = samples.castTo(dtype).rename(outputNames[0])

        return mapOf(outputNames[0] to listOf(output))
    }
}
